### PR TITLE
[Merged by Bors] - feat(Topology): clopen subsets of products of compact spaces are unions of clopen boxes

### PR DESCRIPTION
**Pull request by dagurtomas**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/ClopenBox.lean [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+import Mathlib.Topology.Separation
+/-!
+
+# Clopen subsets in cartesian products
+
+In general, a clopen subset in a cartesian product of topological spaces cannot be written as a
+union of "clopen boxes", i.e. products of clopen subsets of the components (see
+[buzyakovaClopenBox] for counterexamples).
+
+However, when one of the components is compact, a clopen subset can be written as such a union.
```
Comment (jcommelin) [inline Mathlib/Topology/ClopenBox.lean:15]: Minor nitpick:
```suggestion
However, when one of the factors is compact, a clopen subset can be written as such a union.
```

File: Mathlib/Topology/ClopenBox.lean [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+Our argument in `exists_clopen_box` follows the one given in [buzyakovaClopenBox].
+
+We deduce that in a product of compact spaces, a clopen subset is a finite union of clopen boxes,
+and use that to prove that the property of having finitely many clopens is preserved by taking
+cartesian products of compact spaces (this is relevand to the theory of light profinite sets).
```
Comment (jcommelin) [inline Mathlib/Topology/ClopenBox.lean:20]:
```suggestion
cartesian products of compact spaces (this is relevant to the theory of light profinite sets).
```

File: Mathlib/Topology/ClopenBox.lean [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+      have hfi : fI i ∈ (fI '' I).toFinset := by
+        rw [Set.mem_toFinset]
+        exact Set.mem_image_of_mem fI hi
+      use hfi
+    · rw [Set.mem_iUnion] at h
+      obtain ⟨i, hi⟩ := h
+      rw [Set.mem_iUnion] at hi
+      obtain ⟨hi, h⟩ := hi
+      rw [Set.mem_toFinset] at hi
+      obtain ⟨w, hw⟩ := hi
```
Comment (jcommelin) [inline Mathlib/Topology/ClopenBox.lean:167]: Can you golf this by merging the `rw`s into a `simp only`?

File: Mathlib/Topology/ClopenBox.lean [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+In general, a clopen subset in a cartesian product of topological spaces cannot be written as a
+union of "clopen boxes", i.e. products of clopen subsets of the components (see
+[buzyakovaClopenBox] for counterexamples).
+
+However, when one of the components is compact, a clopen subset can be written as such a union.
+Our argument in `exists_clopen_box` follows the one given in [buzyakovaClopenBox].
+
+We deduce that in a product of compact spaces, a clopen subset is a finite union of clopen boxes,
+and use that to prove that the property of having finitely many clopens is preserved by taking
```
Comment (jcommelin) [inline Mathlib/Topology/ClopenBox.lean:19]: Do you mean "finitely many" or "countably many" (which is what you show at the end of the file, and which seems more related to the "light" story)?

File: Mathlib/Topology/ClopenBox.lean [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        left
+        exact ⟨Set.singleton_subset_iff.mpr hx' , subset_refl _⟩
+      refine subset_trans hxt ?_
+      intro ⟨z, w⟩ hz
+      rw [Set.mem_prod] at hz
+      have hz' := hI hz.2
+      rw [Set.mem_iUnion] at hz'
+      obtain ⟨i, hi⟩ := hz'
+      rw [Set.mem_iUnion] at hi
+      obtain ⟨hhi, hi⟩ := hi
```
Comment (jcommelin) [inline Mathlib/Topology/ClopenBox.lean:85]: This looks like another golf opportunity with `simp`?

File: Mathlib/Topology/ClopenBox.lean [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+      apply (this x i.val (hUV (Set.mk_mem_prod hx i.prop))).choose_spec.choose_spec.2.2.2.2
+      rw [Set.mem_prod]
+      refine ⟨?_, hi⟩
+      rw [Set.mem_iInter] at hz
+      have hhz := hz.1 i
+      rw [Set.mem_iInter] at hhz
+      exact hhz hhi
```
Comment (jcommelin) [inline Mathlib/Topology/ClopenBox.lean:92]: And here?

File: Mathlib/Topology/ClopenBox.lean [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    · apply Set.Finite.isOpen_biInter (Set.Finite.ofFinset I (fun _ ↦ Iff.rfl))
+      intro v _
+      exact (this x v.val (hUV (Set.mk_mem_prod hx v.prop))).choose_spec.choose_spec.1
+    · rw [Set.mem_iInter]
+      intro v
+      rw [Set.mem_iInter]
+      intro
```
Comment (jcommelin) [inline Mathlib/Topology/ClopenBox.lean:99]: And here?

**Review by jcommelin — COMMENTED** [review]
The big proof is pretty big... are there natural sublemmas that can be extracted?

**Comment by dagurtomas** [comment]
I extracted the fact that `U` is open as a sublemma

File: Mathlib/Topology/ClopenBox.lean [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+## References
+
+- [buzyakovaClopenBox]: *On clopen sets in Cartesian products*, 2001.
+- [engelking1989]: *General Topology*, 1989.
+
+-/
+
+variable {X Y : Type*} [TopologicalSpace X] [TopologicalSpace Y]
+
+theorem isOpen_set_of_singleton_prod_contained {V : Set Y} (hV : IsCompact V) {W : Set (X × Y)}
```
Comment (urkud) [inline Mathlib/Topology/ClopenBox.lean:31]:
```suggestion
theorem isOpen_setOf_singleton_prod_contained {V : Set Y} (hV : IsCompact V) {W : Set (X × Y)}
```

File: Mathlib/Topology/ClopenBox.lean [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+variable {X Y Z : Type*} [TopologicalSpace X] [TopologicalSpace Y] [TopologicalSpace Z]
+
+lemma isOpen_setOf_mapsTo_isCompact {f : X → Y → Z} (hf : Continuous (uncurry f))
+    {K : Set Y} (hK : IsCompact K) {W : Set Z} (hW : IsOpen W) : IsOpen {x | MapsTo (f x) K W} := by
+  simpa only [mapsTo']
+    using (ContinuousMap.isOpen_gen hK hW).preimage (ContinuousMap.curry ⟨_, hf⟩).continuous
+
+lemma isClosed_setOf_mapsTo {f : X → Y → Z} (hf : ∀ y, Continuous (f · y)) (s : Set Y) {t : Set Z}
+    (ht : IsClosed t) : IsClosed {x | MapsTo (f x) s t} := by
+  simpa only [MapsTo, setOf_forall] using isClosed_biInter fun y _ ↦ ht.preimage (hf y)
```
Comment (urkud) [inline Mathlib/Topology/ClopenBox.lean:42]: This lemma should go to `Topology.Constructions`.

File: Mathlib/Topology/ClopenBox.lean [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+open Function Set Filter TopologicalSpace
+open scoped Topology
+
+variable {X Y Z : Type*} [TopologicalSpace X] [TopologicalSpace Y] [TopologicalSpace Z]
+
+lemma isOpen_setOf_mapsTo_isCompact {f : X → Y → Z} (hf : Continuous (uncurry f))
+    {K : Set Y} (hK : IsCompact K) {W : Set Z} (hW : IsOpen W) : IsOpen {x | MapsTo (f x) K W} := by
+  simpa only [mapsTo']
+    using (ContinuousMap.isOpen_gen hK hW).preimage (ContinuousMap.curry ⟨_, hf⟩).continuous
```
Comment (urkud) [inline Mathlib/Topology/ClopenBox.lean:38]: This lemma should right after `ContinuousMap.curry`.

**Comment by urkud** [comment]
I golfed the proofs using some lemmas that were already in the library. Now we need a review from another maintainer (e.g., I'm not sure if the names I use are too verbose).

**Comment by dagurtomas** [comment]
These changes look great. Thanks @urkud!

**Comment by urkud** [comment]
LGTM but I contributed too much to merge it.
maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**